Nos/Models/NoteParser.swift

```diff
@@ -0,0 +1,105 @@
+//
+//  NoteParser.swift
+//  Nos
+//
+//  Created by Martin Dutra on 17/4/23.
+//
+
+import CoreData
+import Foundation
+import RegexBuilder
+
+/// This struct encapsulates the algorithms that parse notes and the mentions inside the note.
+struct NoteParser {
+
+    /// Parses attributed text generated when composing a note and returns
+    /// the content and tags.
+    func parse(attributedText: AttributedString) -> (String, [[String]]) {
+        cleanLinks(in: attributedText)
+    }
+
+    /// Parses the content and tags stored in a note and returns an attributed text that can be used for displaying
+    /// the note in the UI.
+    func parse(content: String, tags: [[String]], context: NSManagedObjectContext) -> AttributedString {
+        // swiftlint:disable all
+        let regex = /(?:^|\s)#\[(?<index>\d+)\]|(?:^|\s)(?:nostr:)(?<npub>[-a-zA-Z0-9@:%._\+~#=]{2,256})/
+        // swiftlint:enable all
+        let result = content.replacing(regex) { match in
+            let substring = match.0
+            let index = match.1
+            let npub = match.2
+            var prefix = ""
+            let firstCharacter = String(String(substring).prefix(1))
+            if firstCharacter.range(of: #"\s|\r\n|\r|\n"#, options: .regularExpression) != nil {
+                prefix = firstCharacter
+            }
+            var findAndReplaceAuthorReference: (String) -> String = { hex in
+                if let author = try? Author.find(by: hex, context: context) {
```

Comment:
I think we need to wrap this and the fetch of the `Event` further down in `context.perform {`. It isn't safe to use Core Data objects on threads other than the thread associated with context, and although we can see that the only caller of this function right now does wrap its call in `context.perform {` we can't guarantee that will always be true.